File: CP/LeetCode/C/79.c
// 79.c Word Search

#include <stdio.h>
#include <stdbool.h>


bool dfs(char board[6][6], int boardSize, int *boardColSize, char *word, int wordIndex, int row, int col)
{
    if (word[wordIndex] == '\0')
    {
        return true;
    }

    if (row < 0 || row >= boardSize || col < 0 || col >= (*boardColSize) || board[row][col] != word[wordIndex])
    {
        return false;
    }

    char temp = board[row][col];
    board[row][col] = '#';

    bool found = dfs(board, boardSize, boardColSize, word, wordIndex + 1, row + 1, col) ||
                 dfs(board, boardSize, boardColSize, word, wordIndex + 1, row - 1, col) ||
                 dfs(board, boardSize, boardColSize, word, wordIndex + 1, row, col + 1) ||
                 dfs(board, boardSize, boardColSize, word, wordIndex + 1, row, col - 1);

    board[row][col] = temp;
    return found;
}

bool exist(char board[6][6], int boardSize, int *boardColSize, char *word)
{
    for (int i = 0; i < boardSize; i++)
    {
        for (int j = 0; j < (*boardColSize); j++)
        {
            if (dfs(board, boardSize, boardColSize, word, 0, i, j))
            {
                return true;
            }
        }
    }
    return false;
}

int main()
{
    char board[6][6] = {
        {'A', 'B', 'C', 'E'},
        {'S', 'F', 'C', 'S'},
        {'A', 'D', 'E', 'E'},
        {'A', 'B', 'C', 'D'},
        {'A', 'B', 'C', 'E'},
        {'A', 'B', 'C', 'F'}
    };

    int boardColSize = 4;
    char word[] = {'A', 'B', 'C', 'E', '\0'}; 
    bool result = exist(board, 3, &boardColSize, word);

    printf("%s\n", result ? "true" : "false");
    return 0;
}

File: CP/LeetCode/Python/79.py
# 79. Word Search

from __future__ import annotations


class Solution:
    def exist(self, board: list[list[str]], word: str) -> bool:
        if not board or not board[0]:
            return False

        rows, cols = len(board), len(board[0])

        def dfs(r: int, c: int, index: int) -> bool:
            if index == len(word):
                return True
            if r < 0 or r >= rows or c < 0 or c >= cols or board[r][c] == '#':
                return False

            if board[r][c] != word[index]:
                return False

            temp = board[r][c]
            board[r][c] = '#'

            found = dfs(r + 1, c, index + 1) or dfs(r - 1, c, index + 1) or dfs(r, c + 1, index + 1) or dfs(r, c - 1, index + 1)

            board[r][c] = temp
            return found

        for i in range(rows):
            for j in range(cols):
                if board[i][j] == word[0] and dfs(i, j, 0):
                    return True

        return False


if __name__ == "__main__":
    solution = Solution()
    board = [['A', 'B', 'C', 'E'], ['S', 'F', 'C', 'S'], ['A', 'D', 'E', 'E']]
    word = "ABCCED"
    print(solution.exist(board, word))
